core/src/main/scala/pureconfig/DerivedWriters.scala

```diff
@@ -0,0 +1,113 @@
+package pureconfig
+
+import scala.collection.JavaConverters._
+import scala.collection.generic.CanBuildFrom
+import scala.language.higherKinds
+
+import com.typesafe.config._
+import pureconfig.error._
+import shapeless._
+import shapeless.labelled._
+
+/**
+ * Trait containing `ConfigWriter` instances for collection, product and coproduct types.
+ */
+trait DerivedWriters {
+
+  private[pureconfig] trait WrappedConfigWriter[Wrapped, SubRepr] extends ConfigWriter[SubRepr]
+
+  implicit final def hNilConfigWriter[Wrapped]: WrappedConfigWriter[Wrapped, HNil] = new WrappedConfigWriter[Wrapped, HNil] {
+    override def to(t: HNil): ConfigValue = ConfigFactory.parseMap(Map().asJava).root()
+  }
+
+  implicit final def hConsConfigWriter[Wrapped, K <: Symbol, V, T <: HList, U <: HList](
+    implicit
+    key: Witness.Aux[K],
+    vFieldConvert: Lazy[ConfigWriter[V]],
+    tConfigWriter: Lazy[WrappedConfigWriter[Wrapped, T]],
+    hint: ProductHint[Wrapped]): WrappedConfigWriter[Wrapped, FieldType[K, V] :: T] = new WrappedConfigWriter[Wrapped, FieldType[K, V] :: T] {
+
+    override def to(t: FieldType[K, V] :: T): ConfigValue = {
+      val keyStr = hint.configKey(key.value.toString().tail)
+      val rem = tConfigWriter.value.to(t.tail)
+      // TODO check that all keys are unique
```

Comment:
I think we should address old TODOs in another PR because this one is dealing with enough stuff :-)